ouroboros-consensus/docs/report/chapters/storage/overview.tex

```diff
@@ -18,14 +18,74 @@ \section{Components}
 
 Discuss the immutable/volatile split (we reference this section for that).
 
+
 \section{In memory}
 \label{storage:inmemory}
 
 TODO: After we discussed the overview, we should give an overview of everything
 we store in memory in any component, so that we have a better understanding of
 memory usage of the chain DB as a whole.
 
+\subsection{Resource registry}
+\label{storage:resourceregistry}
+
+In order to deal with resource allocation and deallocation, we use the
+abstraction of the \lstinline!ResourceRegistry!. The resource registry is a
+generalization of the \lstinline!bracket! pattern. Using a bracket imposes strict
+rules on the scope of the allocated variables, as once the body of the
+\lstinline!bracket! call finishes, the resources will be deallocated.
+
+On some situations this is too constraining and resources, despite the need to
+ensure they are deallocated, need to live for longer periods or be shared by
+multiple consumers. The registry itself lives in a \lstinline!bracket!-like
+environment, but the resources allocated in it can be opened and closed at any
+point in the lifetime of the registry and will ultimately be closed when the
+registry is being closed if they are still open at that time.
+
+The allocation function for a given resource will run with exceptions
+unconditionally masked, therefore on an uninterruptible fashion. Because of
+this, it is important that such functions are fast. Note that once the
+allocation call finishes, the resource is now managed by the registry and will
+be deallocated in case an exception is thrown.
+
+A resource that is allocated will be returned coupled with its
+\lstinline!ResourceKey! that can be used to release the resource as it holds a
+reference to the registry in which it is allocated.
+
+Special care has to be put when resources are dependent one to another. For
+example, a thread allocated in the registry might hold some mutable reference to
+a file handle that is replaced at certain points during the execution. In such
+cases, the sequence of deallocation must take this into account and deallocate
+the resources in reverse order of dependency. For this purpose, the resource
+registry offers an API to replace a specific resource without altering its
+position in the deallocation sequence.
+
+\subsubsection{Temporary registries}
+\label{storage:temporaryregs}
+
+When some resources are not to be allocated in a registry, one can take
+advantage of resource registries as a temporary container for those
+resources. For this purpose, the \lstinline!WithTempRegistry! is made available. It
```

Comment:
Can we give an example of situations in which we might need this (in consensus)?

Reply:
We currently don't use it anywhere. It was used in the ChainDB initialization but not anymore. But I thought it might be worth mentioning it as it looks like an interesting feature.

Reply:
Outdated, closing